src/include/sof/audio/component.h

```diff
@@ -193,6 +193,10 @@ struct comp_ops {
 	int (*params)(struct comp_dev *dev,
 		      struct sof_ipc_stream_params *params);
 
+	/** get dai hw parameters */
+	int (*dai_get_hw_params)(struct comp_dev *dev,
```

Comment:
I don't know what you mean by "we call this (```dai_get_hw_params()``` function) on volume component". Volume component does not have implemented ```dai_get_hw_params()```, so nothing would happen.
In case above, assuming that selector can have two sinks, firstly in ```pipeline_params()``` we invoke ```pipeline_comp_hw_params()```, which "search" dai component in pipeline, so it goes to DAI1 and fetches its hardware params by ```dai_get_hw_params()``` (which were set during ```dai_set_config()```) then goes to DAI0 and also invokes ```dai_get_hw_params()``` (it would overwrite hardware DAI1 parameters) then goes to remaining components and sets its buffer parameter (by ```buffer_set_hw_params()```) with hardware DAI0 parameters. Then we start second walk on graph with pcm parameters from driver (```pipeline_comp_params()```). Volume and selector components set their sink buffer with pcm params (selector checks whether it can convert "pcm" channels to "hardware" channels). At the end DAI0 and DAI1 in dai_verify() verifies whether parameters in theirs source buffers parameters are "compatible" (in fact, are equal) with hardware parameters. In that case, if DAI0 and DAI1 had the same configuration, no error will appear.
What do you think? Would be grateful for any suggestions.

Reply:
@bkokoszx If the parameters differ there is no way to reconcile them I assume?

Reply:
@paulstelian97 
Yes, in that case (a bit strange selector component usage).

But I think it is easier to consider following topology (```DEMUX```is able to produce streams with different number of channels):
```
[HOST] -> [BUF2] -> [DEMUX] -+-> [BUF0] -> [DAI0]
                             |
                             +-> [BUF1] -> [DAI1]
```
In that case, we invoke ```pipeline_comp_hw_params()```. We go to ```DAI1``` and set ```BUF1``` hardware parameters with ```DAI1``` hardware parameters using:
```
if (calling_buf && !calling_buf->hw_params_configured)
		buffer_set_hw_params(calling_buf, &ppl_data->params->params);
```
Then we go to ```DAI0``` and similarly set ```BUF0``` with ```DAI0``` hw params. Next we come back and set ```BUF2``` with ```DAI0``` hw params.
Subsequently, we start second walk graph with PCM parameters from driver (```pipeline_comp_params()```). ```HOST``` overwrites ```BUF2``` parameters with PCM, as ```HOST``` component is not able to change any parameter. Next, ```DEMUX``` component changes ```BUF0``` and ```BUF1``` with PCM parameters instead of "channels" (we add specific ```BUFF_PARAMS_CHANNELS``` flag to ```comp_verify_params()``` function):
```
comp_verify_params(dev, BUFF_PARAMS_CHANNELS, params, NULL);
```
In effect, ```BUF1``` and ```BUF0``` will have set PCM parameters with unchanged channels parameter set during ```pipeline_comp_hw_params()``` (they could be different).

Reply:
@bkokoszx ok, so I was meaning that I'm seeing ```dai_get_hw_params()``` as a comp_op, meaning that a user can call this on any component. So my question is more about where is it valid to make this call, can you give me an example of this works on your DEMUX pipeline. Does the pipeline only call this on DAI components?

Reply:
So ```pipeline_comp_hw_params()``` is non-tail recursive function (we make some operations after recursive call). It uses ```pipeline_for_each_comp()``` function like like any other ```pipeline_comp_*()``` function. It always starts from host and goes downstream for playback and upstream for capture. Firstly, we go the last component in list. In above ```DEMUX``` topology we reach "last" ```DAI1``` component. When we reach the "last" component we walk down from stack and invoke following part of code:
```
/* Fetch hardware stream parameters from DAI component */
	ret = comp_dai_get_hw_params(current, &ppl_data->params->params);
	if (ret < 0) {
		trace_pipe_error("pipeline_find_dai_comp(): "
				 "comp_dai_get_hw_params() error.");
		return ret;
	}

	/* set buffer parameters */
	if (calling_buf && !calling_buf->hw_params_configured)
		buffer_set_hw_params(calling_buf, &ppl_data->params->params);
```
Therefore:
1. We invoke ```comp_dai_get_hw_params(DAI1, params)``` and fetch ```DAI1``` hardware parameters and set it to ```BUF1``` using:
```
   if (calling_buf && !calling_buf->hw_params_configured)
		buffer_set_hw_params(calling_buf, &ppl_data->params->params);
```
2. We walk down from stack to ```DAI0``` and do the same thing - invoke ```comp_dai_get_hw_params(DAI0, params)``` and fetch ```DAI0``` hardware parameters and set it to ```BUF0```.
3. Then we goes to ```DEMUX``` component. We call ```comp_dai_get_hw_params(DEMUX, params)```, but ```DEMUX``` does not have implemented this function, so nothing happen. Params are still equal to the ```DAI0``` hardware parameters (from the previous step) and we set them to ```BUF2```.
4. The end. :)

Due to your suggestions we can do something like:
```diff
diff --git a/src/audio/pipeline.c b/src/audio/pipeline.c
index 40fb95cd..2684e6f7 100644
--- a/src/audio/pipeline.c
+++ b/src/audio/pipeline.c
@@ -288,11 +288,13 @@ static int pipeline_comp_hw_params(struct comp_dev *current,
                               NULL, dir);

        /* Fetch hardware stream parameters from DAI component */
-       ret = comp_dai_get_hw_params(current, &ppl_data->params->params);
-       if (ret < 0) {
-               trace_pipe_error("pipeline_find_dai_comp(): "
-                                "comp_dai_get_hw_params() error.");
-               return ret;
+       if (current->comp.type == SOF_COMP_DAI) {
+               ret = comp_dai_get_hw_params(current, &ppl_data->params->params);
+               if (ret < 0) {
+                       trace_pipe_error("pipeline_find_dai_comp(): "
+                                        "comp_dai_get_hw_params() error.");
+                       return ret;
+               }
        }

        /* set buffer parameters */
diff --git a/src/include/sof/audio/component.h b/src/include/sof/audio/component.h
index c1ec43f3..0287d5e4 100644
--- a/src/include/sof/audio/component.h
+++ b/src/include/sof/audio/component.h
@@ -430,6 +430,13 @@ static inline int comp_params(struct comp_dev *dev,
 static inline int comp_dai_get_hw_params(struct comp_dev *dev,
                                         struct sof_ipc_stream_params *params)
 {
+       if (dev->comp.type != SOF_COMP_DAI) {
+               trace_comp_error_with_ids(dev,
+                                         "comp_dai_get_hw_params() error: invalid component type"
+                                        );
+               return -EINVAL;
+       }
+
        if (dev->drv->ops.dai_get_hw_params)
                return dev->drv->ops.dai_get_hw_params(dev, params);
        return 0;
```